### PR TITLE
[FEATURE] z* and z' ions for ExD

### LINKED ISSUE #7219: Support different proton abstracted, radicalized, ... c/z ions for ECD
When I was comparing different theoretical spectrum generation methods I encountered vast differences in how they handle c/z ions.

see also https://skyline.ms/announcements/home/support/thread.view?rowId=38133
and
https://www.sciencedirect.com/science/article/pii/S1387380615002249#bib0210

for the background on which c/z ions can be produced and why there is a need for different ones.

### PR DESCRIPTION
**Pull request by jpfeuffer**
## Description

- fixes #7219 
- fixes huge bugs in getMonoWeight(iontype). Affected MRMAssay to be completely wrong
- adds more activation methods
- makes precalculated ion shifts static

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp [hunk 1]
```diff
@@ -471,30 +471,34 @@ START_SECTION(static MSSpectrum generateSpectrum(const Precursor::ActivationMeth
...
+  TEST_REAL_SIMILAR(spec[9].getPosition()[0], 503.2322692);
+  TEST_REAL_SIMILAR(spec[10].getPosition()[0], 578.3085457);
+  // ...
 
   spec.clear(true);
 
   // Test precursor_charge > 2
   spec = TheoreticalSpectrumGenerator::generateSpectrum(Precursor::ActivationMethod::HCID, AASequence::fromString("PEP"), 3);
-  ABORT_IF(spec.size() != 6);
+  TEST_EQUAL(spec.size(), 8);
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:493]: what are the new ion types here?

Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:493]: a-ions right?

Comment (jpfeuffer) [reply src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:493]: Yes should be the a-ions that are more likely in HCID, though still not as common as b/y.
This is debatable if you want them here or not. I thought rather too many than too few.

**Comment by timosachsenberg** [comment]
can you expand a bit on the bug and how you fixed it.

**Comment by jpfeuffer** [comment]
I was trying to compare to other tools by calling getMonoWeight on single residues with Type z-ion.
But it was completely off. Then I saw that all the getMonoWeight(type) functions start from the full AA and add the mass diff for (internal residue to ion type). So you are basically a full water off.
Unless this is weirdly handled outside of the function whenever it is called in the MRM code, it was wrong all the time.

File: src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp (hunk 1 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:493]: They are kind of rare and I usually only observe partial coverage (e.g. just a2,a3). But this is also very much dependent on collision energy and data.
But I think your approach of creating more ions is sensible.
Someone who writes a e.g. search engine probably wants to specify ion types manually anyway and not via this enum. So probably not such an issue...

**Comment by timosachsenberg** [comment]
> I was trying to compare to other tools by calling getMonoWeight on single residues with Type z-ion. But it was completely off. Then I saw that all the getMonoWeight(type) functions start from the full AA and add the mass diff for (internal residue to ion type). So you are basically a full water off. Unless this is weirdly handled outside of the function whenever it is called in the MRM code, it was wrong all the time.

Hmm so you mean if you make a transition for the first prefix/suffix ions in general?
I think they are basically never observed, so I guess no harm done because they are probably not used for transition (and also the reason it was not detected).

**Comment by jpfeuffer** [comment]
Yes correct, I agree but I wonder why the test files changed then.
I don't think they used annotation based on activation_method which is the only other change.

**Comment by timosachsenberg** [comment]
> Yes correct, I agree but I wonder why the test files changed then. I don't think they used annotation based on activation_method which is the only other change.

maybe they take all HCD ions? and you get different results when a-ions are included?

**Comment by jpfeuffer** [comment]
Maybe, but I think this feature (of specifying activation methods for TSG) was pretty new. And I didn't see the TSG there. Maybe I missed it.

**Comment by timosachsenberg** [comment]
Another idea: maybe they mess around with element/isotopes somewhere (e.g. 13C/15N labeling by changing ElementsDB) and this is only picked up if it is not static but calculated on the fly?

**Comment by jpfeuffer** [comment]
Good idea, I quickly tried but it is not. It's the change in mass calculation.

**Comment by jpfeuffer** [comment]
Those tests are even failing on CI. They pass here, that's why I adapted the tests. Something else is fishy.
And I have zero idea (and time to investigate) why that weird cwl file converter test fails.

**Comment by jpfeuffer** [comment]
Probably someone without a silicon Mac needs to finish this, otherwise we cannot rule out std lib/float calc/sorting differences

**Comment by jpfeuffer** [comment]
When tests are suffixed with boost58, this is not a good sign anyways...

**Comment by timosachsenberg** [comment]
Yeah I guess there are some pointers about that in MRMAssay
```
#include <boost/random/mersenne_twister.hpp>
#include <boost/random/uniform_int.hpp>
#include <boost/random/variate_generator.hpp>
#include <boost/unordered_map.hpp> // cannot remove this since tests fail otherwise
```

and

```
    /**
      @brief Generate decoy sequences

      Generate decoy sequences for IPF algorithm which share peptidoform
      properties with targets.

      @details Used internally by the IPF algorithm, see MRMAssay::uisTransitions()

    */
    void generateDecoySequences_(const SequenceMapT& TargetSequenceMap,
                                 boost::unordered_map<String, String>& DecoySequenceMap,
                                 int shuffle_seed);
```

**Comment by timosachsenberg** [comment]
I think everything points to the use of boost for unordered map and random number generation. I think we could disable the uis tests until someone makes random number generators deterministic.

**Comment by jpfeuffer** [comment]
You can maybe try adding a regex to Fuzzydiff for UIs decoy transitions.

Don't we have a "deterministic" rng implemented somewhere?

**Comment by jpfeuffer** [comment]
Well at least for shuffling:
https://github.com/OpenMS/OpenMS/blob/cd6baf09074c3baa4899db855e6a187f3dd1e76f/src/openms/include/OpenMS/MATH/MISC/MathFunctions.h#L381C14-L381C14

**Comment by timosachsenberg** [comment]
Feel free to remove the tests. otherwise, not sure when I get to look at it.

**Comment by timosachsenberg** [comment]
/ceph/ibmi/abi/projects/sachsenb/OpenMS/src/openms/source/ANALYSIS/OPENSWATH/TransitionTSVFile.cpp:1302:16: warning: enumeration value ‘Zp1Ion’ not handled in switch [-Wswitch]
 1302 |         switch (int_it.iontype)

**Comment by jpfeuffer** [comment]
thanks, I indeed overlooked that

**Pull request closed without merging**